core/src/main/java/org/apache/iceberg/rest/http/ErrorHandlers.java

```diff
@@ -0,0 +1,119 @@
+
+      switch (errorResponse.getCode()) {
+        case HttpStatus.SC_NOT_FOUND:
+          // TODO: Exception handling here could be better
+          // some methods can result in different resource not found exceptions, so here we need to
+          // differentiate between 404 for non-existent Namespace/Database by looking at X-Iceberg-Exception header.
+          if (NoSuchNamespaceException.class.getSimpleName().equals(responseException)) {
+            throw new NoSuchNamespaceException("Resource not found: %s", responseBody);
+          } else {
+            throw new NoSuchTableException("Resource not found: %s", responseBody);
+          }
+        case HttpStatus.SC_CONFLICT:
+          throw new AlreadyExistsException("Already exists: %s", responseBody);
+        case HttpStatus.SC_FORBIDDEN:
+        case HttpStatus.SC_UNAUTHORIZED:
+          throw new AuthorizationDeniedException("Not Authorized: %s", responseBody);
+        default:
+          throw new RestException("Unknown error: %s", errorResponse);
+      }
+    };
+  }
+
+  /**
+   * Request error handlers specifically for CRUD ops on databases / namespaces.
+   */
+  public static Consumer<CloseableHttpResponse> databaseErrorHandler() {
+    return (errorResponse) -> {
+      String responseBody = getResponseBody(errorResponse);
+
+      switch (errorResponse.getCode()) {
+        case HttpStatus.SC_NOT_FOUND:
+          throw new NoSuchNamespaceException("Namespace not found: %s", responseBody);
+        case HttpStatus.SC_CONFLICT:
+          throw new AlreadyExistsException("Already exists: %s", responseBody);
+        case HttpStatus.SC_FORBIDDEN:
+        case HttpStatus.SC_UNAUTHORIZED:
+          throw new AuthorizationDeniedException("Not Authorized: %s", responseBody);
+        default:
+          throw new RestException("Unknown error: %s", errorResponse);
+      }
+    };
+  }
+
+  static String getResponseBody(CloseableHttpResponse response) {
+    String responseBody = "Non-parseable Response Body";
+    try {
+      responseBody = EntityUtils.toString(response.getEntity());
+    } catch (Exception e) {
+      LOG.error("Encountered an exception getting response body", e);
+    }
+
+    return responseBody;
+  }
+
+  static String getIcebergExceptionHeader(CloseableHttpResponse response) {
+    String icebergException = null;
+    try {
+      // TODO - Extract more specific exception from a header or from the response body?
+      //        Some servers/proxies will strip headers that aren't in the whitelist so possibly headers
+      //        aren't the way to go (though that's typically more on the Request end than the Response end).
+      icebergException = response.getHeader("X-Iceberg-Exception").getValue();
+    } catch (Exception e) {
+      // TODO - Better error message and handling. Will be refactoring this anyway.
+      LOG.error("Encountered an error when getting the X-Iceberg-Exception header", e);
```

Comment:
Instead of using a header, what about having a way to encode an exception in the JSON payload? That seems to be the approach outlined here: https://www.baeldung.com/rest-api-error-handling-best-practices

Reply:
Agreed. That's more in line with how I've previously done it as well. The Facebook example is most in line with what I've previously used, where there's a dedicted `error` payload in the response body and then possibly even further error codes beyond just the normal 4xx, and those error codes map to what happened.

```
{
    "error": {
        "message": "Missing Bearer header.",
        "type": "OAuthException",
        "code": 40101,  // We document what 401xx error codes correspond to as part of the spec
        "trace_id": "AWswcVwbcqfgrSgjG80MtqJ"
    },
    data: {
         // Successful response payload 
    }
}
```

I will refactor to use an approach like that.

Reply:
+1. I'm not sure that we need error codes, but we can always add them if people find them useful.